python/libucx/libucx/load.py

```diff
@@ -33,19 +33,53 @@
 ]
 
 
+# Loading with RTLD_LOCAL adds the library itself to the loader's
+# loaded library cache without loading any symbols into the global
+# namespace. This allows libraries that express a dependency on
+# a library to be loaded later and successfully satisfy that dependency
+# without polluting the global symbol table with symbols from
+# that library that could conflict with symbols from other DSOs.
+PREFERRED_LOAD_FLAG = ctypes.RTLD_GLOBAL
```

Comment:
It seems like maybe `RTLD_GLOBAL` is required for these libraries? Maybe because they don't have the correct `DT_NEEDED` entries?

https://github.com/rapidsai/ucx-wheels/blob/ff1946193342bdc75c082c3d6aa153d8cdd400a9/python/libucx/libucx/load.py#L19-L26

Using `RTLD_LOCAL`, importing failed like this:

```text
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/pyenv/versions/3.11.10/lib/python3.11/site-packages/libucx/load.py", line 81, in load_library
    libucx_lib = _load_wheel_installation(lib)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/pyenv/versions/3.11.10/lib/python3.11/site-packages/libucx/load.py", line 57, in _load_wheel_installation
    return ctypes.CDLL(lib, PREFERRED_LOAD_FLAG)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/pyenv/versions/3.11.10/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: /pyenv/versions/3.11.10/lib/python3.11/site-packages/libucx/lib/libucm.so: undefined symbol: ucs_load_modules
```

([build link](https://github.com/rapidsai/ucx-wheels/actions/runs/11845661052/job/33011919914))

Switching to `RTLD_GLOBAL` in https://github.com/rapidsai/ucx-wheels/pull/13/commits/285f51407853bf0292b4b7107a8c15b46a975717, loading the libraries succeeded (https://github.com/rapidsai/ucx-wheels/actions/runs/11847604113).

Reply:
We should definitely not use `RTLD_GLOBAL`. `libucm.so` should include libucs as a DT_NEEDED entry.

Reply:
I'll let @vyasr comment on that. He wrote that comment I linked and will remember the context for it.

Reply:
I don't understand the comment at the top of this file:
```python
# IMPORTANT: The load order here matters! libucm.so depends on symbols in libucs.so, but
# it does not express this via a DT_NEEDED entry, presumably because libucs.so also has
# a dependency on libucm.so and the libraries are attempting to avoid a circular
# dependency.
```

If `libucm.so` needs `libucs.so`, it should have a `DT_NEEDED` entry. If `libucs.so` depends on symbols from `libucm.so`, then how can it be loaded before `libucm.so`?

Reply:
I'll need to reproduce this to demonstrate the issue, but basically what I recall observing is that if you inspected the symbols in libucs.so it had some unknown symbols that were defined in libucm.so, but the reverse was also true because libucm.so had unknown symbols that were defined in libucs.so. However, only one of these had a DT_NEEDED entry for the other. Therefore, the load order mattered because you could load the one that had the DT_NEEDED entry and with the right RPATH set it would load the other library (which would be loaded successfully because it saw the symbols from the first library) and then everything would be satisfied, but if loaded the other library first then it would simply fail to resolve all needed symbols.

If your response is "that doesn't make sense"... you're right, it doesn't. I don't know how or why it was set up this way. The circular dependency thought was my hypothesis. I'll set up a reproducer and we can inspect it. Worst case scenario we could simply use patchelf to add the missing DT_NEEDED entry to force loads to work correctly.

Reply:
We discussed this offline and confirmed my hypothesis. Circular DT_NEEDED entries are indeed UB for the loader, so ucx is intentionally being built in such a way as to avoid them by using the `--no-as-needed` linker flag. Fortunately, this means that to avoid the problem we can simply avoid loading `libucm.so` explicitly altogether and just load `libucs.so`, which will bring it in transitively for us.

Reply:
Pushed the updated fix here.